## Exploratory Data Analysis

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

pd.set_option('display.max_colwidth', None)


In [2]:
# Dataset
dataset = pd.read_csv("/Users/jaskirat/University_at_Buffalo/courses/statistical_learning/project/data/amazon_transaction_data_100K.csv")
dataset.head()

EVENT_LABEL       EVENT_TIMESTAMP       LABEL_TIMESTAMP  \
0            0  2023-02-14T02:27:41Z  2023-02-14T02:27:41Z   
1            0  2023-01-13T01:30:22Z  2023-01-13T01:30:22Z   
2            0  2023-02-20T04:12:39Z  2023-02-20T04:12:39Z   
3            0  2022-12-13T09:06:33Z  2022-12-13T09:06:33Z   
4            0  2023-03-04T09:48:57Z  2023-03-04T09:48:57Z   

                            EVENT_ID ENTITY_TYPE    ENTITY_ID  card_bin  \
0  100005a527cbcd5de0c4c878288912017    customer  634-80-3034    259703   
1  1000091e85eb1cd32ec4b8afcf818fea5    customer  815-02-7980    380256   
2  10000a7862d0ca68257a20149111ec1b4    customer  287-97-0013    471448   
3  1000130fddbb2383646840294dcb7f6fa    customer  819-30-8170    343617   
4  1000230990f537b8e89786821a3abf10e    customer  285-06-1069    353899   

  customer_name                       billing_street billing_city  ...  \
0         Ricky  8247 Christopher Mountains Apt. 462   Paint Lick  ...   
1      Kimberly                       945 Lowe Vista   Providence  ...   
2        Ronald        11307 Munoz Mission Suite 083    Arlington  ...   
3       Jessica              90815 Brandon Junctions    San Diego  ...   
4        Robert             63165 Natasha Extensions     Mayfield  ...   

  billing_country                    customer_job       ip_address  \
0              US           IT sales professional    120.79.45.214   
1              US           Aeronautical engineer    212.42.56.229   
2              US                Ambulance person  164.238.228.201   
3              US            Nurse, mental health  210.108.230.215   
4              US  Teaching laboratory technician  189.103.115.129   

             customer_email      billing_phone  \
0   woodardbrenda@gmail.com       786-385-2594   
1  scottdalton@robinson.biz  266-305-0756-3308   
2        pjohnson@cruz.info    +1-068-919-8279   
3     progers@hansen-yu.com    +1-115-473-9557   
4   reevesmichael@gmail.com    +1-057-636-3058   

                                                                                                                                    user_agent  \
0                                                                              Mozilla/5.0 (compatible; MSIE 5.0; Windows NT 5.2; Trident/5.1)   
1           Mozilla/5.0 (Macintosh; Intel Mac OS X 10_6_4 rv:6.0; cmn-TW) AppleWebKit/531.26.2 (KHTML, like Gecko) Version/5.0 Safari/531.26.2   
2  Mozilla/5.0 (iPhone; CPU iPhone OS 10_3_4 like Mac OS X) AppleWebKit/533.2 (KHTML, like Gecko) FxiOS/18.5b4414.0 Mobile/13C311 Safari/533.2   
3                                                                                  Mozilla/5.0 (compatible; MSIE 6.0; Windows CE; Trident/4.1)   
4        Mozilla/5.0 (Macintosh; U; PPC Mac OS X 10_11_8 rv:6.0; cs-CZ) AppleWebKit/531.15.7 (KHTML, like Gecko) Version/4.0.2 Safari/531.15.7   

  product_category order_price payment_currency                   merchant  
0      grocery_net       95.77              ARS  fraud_Runolfsson and Sons  
1        kids_pets       21.88              BND           fraud_Larkin Ltd  
2     shopping_pos        9.62              NAD           fraud_Metz-Boehm  
3             home       36.96              LTL      fraud_Quitzon-Goyette  
4     shopping_pos        2.08              GIP      fraud_Turner and Sons  

[5 rows x 24 columns]

In [3]:
len(dataset)

116705

In [4]:
dataset["EVENT_LABEL"].value_counts()

0    104804
1     11901
Name: EVENT_LABEL, dtype: int64

In [5]:
dataset.dtypes

EVENT_LABEL            int64
EVENT_TIMESTAMP       object
LABEL_TIMESTAMP       object
EVENT_ID              object
ENTITY_TYPE           object
ENTITY_ID             object
card_bin               int64
customer_name         object
billing_street        object
billing_city          object
billing_state         object
billing_zip            int64
billing_latitude     float64
billing_longitude    float64
billing_country       object
customer_job          object
ip_address            object
customer_email        object
billing_phone         object
user_agent            object
product_category      object
order_price          float64
payment_currency      object
merchant              object
dtype: object

In [6]:
# Filter columns
columns_to_keep = [
    "EVENT_LABEL",
    "LABEL_TIMESTAMP",
    "EVENT_ID",
    "ENTITY_ID",
    "customer_job",
    "user_agent",
    "product_category",
    "order_price",
    "merchant"
]
filtered_dataset = dataset[columns_to_keep]

In [7]:
filtered_dataset.dtypes

EVENT_LABEL           int64
LABEL_TIMESTAMP      object
EVENT_ID             object
ENTITY_ID            object
customer_job         object
user_agent           object
product_category     object
order_price         float64
merchant             object
dtype: object

In [8]:
# Changing all column names to lower case
filtered_dataset = filtered_dataset.rename(columns={
    "EVENT_LABEL": "event_label", 
    "LABEL_TIMESTAMP": "label_timestamp",
    "EVENT_ID": "event_id",
    "ENTITY_ID": "entity_id"
})
filtered_dataset.dtypes


event_label           int64
label_timestamp      object
event_id             object
entity_id            object
customer_job         object
user_agent           object
product_category     object
order_price         float64
merchant             object
dtype: object

In [9]:
# Converting label_timestamp column from string to timestamp
filtered_dataset["label_timestamp"] = pd.to_datetime(
    filtered_dataset["label_timestamp"], 
    infer_datetime_format=True
)


In [10]:
filtered_dataset.head()

event_label     label_timestamp                           event_id  \
0            0 2023-02-14 02:27:41  100005a527cbcd5de0c4c878288912017   
1            0 2023-01-13 01:30:22  1000091e85eb1cd32ec4b8afcf818fea5   
2            0 2023-02-20 04:12:39  10000a7862d0ca68257a20149111ec1b4   
3            0 2022-12-13 09:06:33  1000130fddbb2383646840294dcb7f6fa   
4            0 2023-03-04 09:48:57  1000230990f537b8e89786821a3abf10e   

     entity_id                    customer_job  \
0  634-80-3034           IT sales professional   
1  815-02-7980           Aeronautical engineer   
2  287-97-0013                Ambulance person   
3  819-30-8170            Nurse, mental health   
4  285-06-1069  Teaching laboratory technician   

                                                                                                                                    user_agent  \
0                                                                              Mozilla/5.0 (compatible; MSIE 5.0; Windows NT 5.2; Trident/5.1)   
1           Mozilla/5.0 (Macintosh; Intel Mac OS X 10_6_4 rv:6.0; cmn-TW) AppleWebKit/531.26.2 (KHTML, like Gecko) Version/5.0 Safari/531.26.2   
2  Mozilla/5.0 (iPhone; CPU iPhone OS 10_3_4 like Mac OS X) AppleWebKit/533.2 (KHTML, like Gecko) FxiOS/18.5b4414.0 Mobile/13C311 Safari/533.2   
3                                                                                  Mozilla/5.0 (compatible; MSIE 6.0; Windows CE; Trident/4.1)   
4        Mozilla/5.0 (Macintosh; U; PPC Mac OS X 10_11_8 rv:6.0; cs-CZ) AppleWebKit/531.15.7 (KHTML, like Gecko) Version/4.0.2 Safari/531.15.7   

  product_category  order_price                   merchant  
0      grocery_net        95.77  fraud_Runolfsson and Sons  
1        kids_pets        21.88           fraud_Larkin Ltd  
2     shopping_pos         9.62           fraud_Metz-Boehm  
3             home        36.96      fraud_Quitzon-Goyette  
4     shopping_pos         2.08      fraud_Turner and Sons

In [11]:
filtered_dataset["hour_of_day"] = filtered_dataset.label_timestamp.apply(lambda x: x.hour)
filtered_dataset["minutes_of_day"] = filtered_dataset.label_timestamp.apply(lambda x: (x.hour * 60) + x.minute)
filtered_dataset.head()

event_label     label_timestamp                           event_id  \
0            0 2023-02-14 02:27:41  100005a527cbcd5de0c4c878288912017   
1            0 2023-01-13 01:30:22  1000091e85eb1cd32ec4b8afcf818fea5   
2            0 2023-02-20 04:12:39  10000a7862d0ca68257a20149111ec1b4   
3            0 2022-12-13 09:06:33  1000130fddbb2383646840294dcb7f6fa   
4            0 2023-03-04 09:48:57  1000230990f537b8e89786821a3abf10e   

     entity_id                    customer_job  \
0  634-80-3034           IT sales professional   
1  815-02-7980           Aeronautical engineer   
2  287-97-0013                Ambulance person   
3  819-30-8170            Nurse, mental health   
4  285-06-1069  Teaching laboratory technician   

                                                                                                                                    user_agent  \
0                                                                              Mozilla/5.0 (compatible; MSIE 5.0; Windows NT 5.2; Trident/5.1)   
1           Mozilla/5.0 (Macintosh; Intel Mac OS X 10_6_4 rv:6.0; cmn-TW) AppleWebKit/531.26.2 (KHTML, like Gecko) Version/5.0 Safari/531.26.2   
2  Mozilla/5.0 (iPhone; CPU iPhone OS 10_3_4 like Mac OS X) AppleWebKit/533.2 (KHTML, like Gecko) FxiOS/18.5b4414.0 Mobile/13C311 Safari/533.2   
3                                                                                  Mozilla/5.0 (compatible; MSIE 6.0; Windows CE; Trident/4.1)   
4        Mozilla/5.0 (Macintosh; U; PPC Mac OS X 10_11_8 rv:6.0; cs-CZ) AppleWebKit/531.15.7 (KHTML, like Gecko) Version/4.0.2 Safari/531.15.7   

  product_category  order_price                   merchant  hour_of_day  \
0      grocery_net        95.77  fraud_Runolfsson and Sons            2   
1        kids_pets        21.88           fraud_Larkin Ltd            1   
2     shopping_pos         9.62           fraud_Metz-Boehm            4   
3             home        36.96      fraud_Quitzon-Goyette            9   
4     shopping_pos         2.08      fraud_Turner and Sons            9   

   minutes_of_day  
0             147  
1              90  
2             252  
3             546  
4             588

In [12]:
filtered_dataset.describe()

event_label    order_price    hour_of_day  minutes_of_day
count  116705.000000  116705.000000  116705.000000   116705.000000
mean        0.101975     114.683963      11.545846      722.190969
std         0.302617     229.441548       6.894782      413.967196
min         0.000000       1.000000       0.000000        0.000000
25%         0.000000      11.730000       6.000000      367.000000
50%         0.000000      53.220000      12.000000      723.000000
75%         0.000000      97.000000      17.000000     1079.000000
max         1.000000   22059.750000      23.000000     1439.000000

### Creating time windows

#### Morning -> 6 - 11:59
#### Afternoon -> 12 - 17:59
#### Evening -> 18 - 23:59
#### Midnight -> 0 - 5:59

#### Minutes for midnight -> 0 <= midnight <= 359
#### Minutes for morning -> 359 < morning <= 719
#### Minutes for afternoon -> 719 < afternoon <= 1079
#### Minutes for evening -> 1079 < evening <= 1439

In [13]:
filtered_dataset.dtypes

event_label                  int64
label_timestamp     datetime64[ns]
event_id                    object
entity_id                   object
customer_job                object
user_agent                  object
product_category            object
order_price                float64
merchant                    object
hour_of_day                  int64
minutes_of_day               int64
dtype: object

In [14]:
def get_time_of_day(dataset_row):
    if dataset_row["minutes_of_day"] >= 0 and dataset_row["minutes_of_day"] <= 359:
        time_of_day = "midnight"
    
    elif dataset_row["minutes_of_day"] >= 360 and dataset_row["minutes_of_day"] <= 719:
        time_of_day = "morning"
    
    elif dataset_row["minutes_of_day"] >= 720 and dataset_row["minutes_of_day"] <= 1079:
        time_of_day = "afternoon"
    
    else:
        time_of_day = "evening"
    
    return time_of_day

filtered_dataset["time_of_day"] = filtered_dataset.apply(get_time_of_day, axis=1)
filtered_dataset.dtypes

event_label                  int64
label_timestamp     datetime64[ns]
event_id                    object
entity_id                   object
customer_job                object
user_agent                  object
product_category            object
order_price                float64
merchant                    object
hour_of_day                  int64
minutes_of_day               int64
time_of_day                 object
dtype: object

In [15]:
filtered_dataset.head()

event_label     label_timestamp                           event_id  \
0            0 2023-02-14 02:27:41  100005a527cbcd5de0c4c878288912017   
1            0 2023-01-13 01:30:22  1000091e85eb1cd32ec4b8afcf818fea5   
2            0 2023-02-20 04:12:39  10000a7862d0ca68257a20149111ec1b4   
3            0 2022-12-13 09:06:33  1000130fddbb2383646840294dcb7f6fa   
4            0 2023-03-04 09:48:57  1000230990f537b8e89786821a3abf10e   

     entity_id                    customer_job  \
0  634-80-3034           IT sales professional   
1  815-02-7980           Aeronautical engineer   
2  287-97-0013                Ambulance person   
3  819-30-8170            Nurse, mental health   
4  285-06-1069  Teaching laboratory technician   

                                                                                                                                    user_agent  \
0                                                                              Mozilla/5.0 (compatible; MSIE 5.0; Windows NT 5.2; Trident/5.1)   
1           Mozilla/5.0 (Macintosh; Intel Mac OS X 10_6_4 rv:6.0; cmn-TW) AppleWebKit/531.26.2 (KHTML, like Gecko) Version/5.0 Safari/531.26.2   
2  Mozilla/5.0 (iPhone; CPU iPhone OS 10_3_4 like Mac OS X) AppleWebKit/533.2 (KHTML, like Gecko) FxiOS/18.5b4414.0 Mobile/13C311 Safari/533.2   
3                                                                                  Mozilla/5.0 (compatible; MSIE 6.0; Windows CE; Trident/4.1)   
4        Mozilla/5.0 (Macintosh; U; PPC Mac OS X 10_11_8 rv:6.0; cs-CZ) AppleWebKit/531.15.7 (KHTML, like Gecko) Version/4.0.2 Safari/531.15.7   

  product_category  order_price                   merchant  hour_of_day  \
0      grocery_net        95.77  fraud_Runolfsson and Sons            2   
1        kids_pets        21.88           fraud_Larkin Ltd            1   
2     shopping_pos         9.62           fraud_Metz-Boehm            4   
3             home        36.96      fraud_Quitzon-Goyette            9   
4     shopping_pos         2.08      fraud_Turner and Sons            9   

   minutes_of_day time_of_day  
0             147    midnight  
1              90    midnight  
2             252    midnight  
3             546     morning  
4             588     morning

In [16]:
filtered_dataset["user_agent"].unique()[0:100]

array(['Mozilla/5.0 (compatible; MSIE 5.0; Windows NT 5.2; Trident/5.1)',
       'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_6_4 rv:6.0; cmn-TW) AppleWebKit/531.26.2 (KHTML, like Gecko) Version/5.0 Safari/531.26.2',
       'Mozilla/5.0 (iPhone; CPU iPhone OS 10_3_4 like Mac OS X) AppleWebKit/533.2 (KHTML, like Gecko) FxiOS/18.5b4414.0 Mobile/13C311 Safari/533.2',
       'Mozilla/5.0 (compatible; MSIE 6.0; Windows CE; Trident/4.1)',
       'Mozilla/5.0 (Macintosh; U; PPC Mac OS X 10_11_8 rv:6.0; cs-CZ) AppleWebKit/531.15.7 (KHTML, like Gecko) Version/4.0.2 Safari/531.15.7',
       'Mozilla/5.0 (Android 4.1.1; Mobile; rv:42.0) Gecko/42.0 Firefox/42.0',
       'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_1; rv:1.9.2.20) Gecko/2012-02-29 04:55:02 Firefox/3.8',
       'Opera/8.73.(Windows NT 4.0; af-ZA) Presto/2.9.165 Version/11.00',
       'Mozilla/5.0 (compatible; MSIE 7.0; Windows CE; Trident/4.1)',
       'Opera/9.15.(X11; Linux i686; fo-FO) Presto/2.9.190 Version/12.00',
       'Mozilla

In [17]:
"""
Different types of browsers
1. Chrome
2. Safari
3. Mozilla
4. Opera
5. Firefox
6. Google
7. Presto
8. Trident
"""

def get_browser(dataset_row):
    if "chrome" in dataset_row["user_agent"].lower().split("/", 1)[1]:
        device = "chrome"
    elif "safari" in dataset_row["user_agent"].lower().split("/", 1)[1]:
        device = "safari"
    elif "firefox" in dataset_row["user_agent"].lower().split("/", 1)[1]:
        device = "firefox"
    elif "opera" in dataset_row["user_agent"].lower().split("/", 1)[1]:
        device = "opera"
    elif "google" in dataset_row["user_agent"].lower().split("/", 1)[1]:
        device = "chrome"
    elif "mozilla" in dataset_row["user_agent"].lower().split("/", 1)[1]:
        device = "firefox"
    elif "trident" in dataset_row["user_agent"].lower().split("/", 1)[1]:
        device = "trident"
    elif "presto" in dataset_row["user_agent"].lower().split("/", 1)[1]:
        device = "presto"
    else:
        device = "not_known"
        
    return device


filtered_dataset["browser"] = filtered_dataset.apply(get_browser, axis=1)
# ["user_agent"].str.split("/", 1).str[1]
filtered_dataset.dtypes


event_label                  int64
label_timestamp     datetime64[ns]
event_id                    object
entity_id                   object
customer_job                object
user_agent                  object
product_category            object
order_price                float64
merchant                    object
hour_of_day                  int64
minutes_of_day               int64
time_of_day                 object
browser                     object
dtype: object

In [18]:
filtered_dataset["browser"].value_counts()

safari     31873
presto     23744
trident    23056
chrome     19035
firefox    18997
Name: browser, dtype: int64

In [19]:
filtered_dataset["browser"].unique()

array(['trident', 'safari', 'firefox', 'presto', 'chrome'], dtype=object)

In [20]:
"""
Different types of devices
1. Windows
2. Mac
3. Linux
4. Iphone
5. Android
6. Ipad
"""

def get_device(dataset_row):
    if "windows" in dataset_row["user_agent"].lower():
        device = "windows"
    elif "macintosh" in dataset_row["user_agent"].lower():
        device = "mac"
    elif "linux" in dataset_row["user_agent"].lower():
        device = "linux"
    elif "iphone" in dataset_row["user_agent"].lower():
        device = "iphone"
    elif "android" in dataset_row["user_agent"].lower():
        device = "android"
    elif "ipad" in dataset_row["user_agent"].lower():
        device = "ipad"
    else:
        device = "not_known"
        
    return device


filtered_dataset["device"] = filtered_dataset.apply(get_device, axis=1)
filtered_dataset.dtypes

event_label                  int64
label_timestamp     datetime64[ns]
event_id                    object
entity_id                   object
customer_job                object
user_agent                  object
product_category            object
order_price                float64
merchant                    object
hour_of_day                  int64
minutes_of_day               int64
time_of_day                 object
browser                     object
device                      object
dtype: object

In [21]:
print(filtered_dataset["device"].unique())

['windows' 'mac' 'iphone' 'android' 'linux' 'ipad']


In [22]:
filtered_dataset["device"].value_counts()

windows    52514
linux      25813
mac        16667
iphone     12408
android     4876
ipad        4427
Name: device, dtype: int64

In [33]:
list(filtered_dataset["customer_job"].unique())

['IT sales professional',
 'Aeronautical engineer',
 'Ambulance person',
 'Nurse, mental health',
 'Teaching laboratory technician',
 'Retail merchandiser',
 'Commercial/residential surveyor',
 'Geologist, wellsite',
 'Tourist information centre manager',
 'Art gallery manager',
 'Research scientist (life sciences)',
 'Corporate treasurer',
 'Passenger transport manager',
 'Theatre director',
 'Archaeologist',
 'Writer',
 'Health service manager',
 'Dramatherapist',
 'IT trainer',
 'Fine artist',
 'Astronomer',
 'Sales executive',
 'Paramedic',
 'Minerals surveyor',
 'Geneticist, molecular',
 'Multimedia programmer',
 'Radiographer, diagnostic',
 'Research scientist (maths)',
 'Licensed conveyancer',
 'Psychologist, prison and probation services',
 'Retail manager',
 'Veterinary surgeon',
 'Nurse, adult',
 'Maintenance engineer',
 'Civil engineer, consulting',
 'Insurance risk surveyor',
 'Restaurant manager',
 'Mining engineer',
 'Physiological scientist',
 'Ceramics designer',
 'Spor

In [113]:
# software -> IT, Technical, network, software engineer, Data Scientist, site engineer, colour technologist, Database, developer
# other_engineering -> 
# scientist -> scientist, researcher, research
# teacher -> teacher, education, tutor
# airline -> airline, pilot
# medical -> biomedical, hospital, doctor, health, mental, psychotherapist, Chiropractor, dentist, Pharmacologist, Radiation, Pharmacist, ecologist, Psychologist, Therapist, Immunologist, Clinical 
# marketing -> market researcher, multimedia
# banking -> broker, insurance, Retail banker
# civil -> diplomatic, civil, Armed, government
# executive -> chief operating, Chief Strategy
# media -> advertising, editorial
# nature -> nature conservation, nature
# trading -> Trade union, Bonds trader
# lawyer -> Lawyer
# advertising -> journalist



"""
Different types of Customer Jobs
1. Engineer
"""
software = [
    "IT", 
    "technical", 
    "network engineer", 
    "software engineer", 
    "data scientist", 
    "site engineer", 
    "colour technologist", 
    "database", 
    "developer",
    "systems analyst",
    "programmer",
    "data",
    "it consultant",
    "intelligence analyst"
]

mathematician = [
    "stats",
    "statistician",
    "mathematics"
]

other_engineering = [
    "engineer",
]

scientist = [
    "scientist", 
    "researcher", 
    "research",
    "science writer", 
    "scientific", 
    "laboratory",
    "astronomer"

]

teaching = [
    "teacher", 
    "education", 
    "tutor",
    "librarian",
    "academic",
    "professor",
    "learning mentor"
]

airline = [
    "airline", 
    "pilot",
    "cabin crew"
]

medical = [
    "biomedical", 
    "hospital", 
    "doctor", 
    "health", 
    "mental", 
    "psychotherapist", 
    "chiropractor", 
    "dentist", 
    "pharmacologist", 
    "radiation", 
    "pharmacist", 
    "ecologist", 
    "psychologist", 
    "therapist", 
    "immunologist", 
    "clinical",
    "medicine",
    "nurse",
    "orthoptist",
    "oncologist",
    "ophthalmologist",
    "toxicologist",
    "neurosurgeon",
    "homeopath",
    "meteorologist",
    "psychiatrist",
    "haematologist",
    "ergonomist",
    "optometrist",
    "geochemist",
    "pathologist",
    "metallurgist",
    "dietitian",
    "chemist",
    "chiropodist",
    "surgeon",
    "microbiologist",
    "ambulance",
    "diagnostic",
    "paramedic",
    "acupuncturist",
    "medical physicist",
    "medical",
    "midwife",
    "herbalist",
    "osteopath",
    "acupuncturist"
]

marketing = [
    "market researcher", 
    "multimedia",
    "marketing"
]

nature = [
    "nature conservation", 
    "nature"
]

trader = [
    "trade union", 
    "bonds trader",
    "trading",
    "trader",
    "dealer"
]

lawyer = [
    "lawyer", 
    "trade mark attorney", 
    "attorney"
]

journalism = [
    "journalist",
    "journalist",
    "journalism"
]

finance = [
    "broker",
    "insurance",
    "retail banker",
    "banker",
    "banking",
    "financial",
    "pension scheme"
]

civil = [
    "diplomatic", 
    "civil", 
    "armed", 
    "government",
    "prison",
    "tax professional",
    "tax inspector",
    "immigration officer",
    "information officer",
    "personnel officer"
]

executive = [
    "chief operating", 
    "chief strategy",
    "chief marketing",
    "chief of staff",
    "chief executive",
    "chief financial",
    "chief technology",
    "executive"
]

media = [
    "advertising", 
    "editorial",
    "publishing rights manager"
]

politician = [
    "politics",
    "politician"
]

sports = [
    "sports",
    "sport",
    "fitness"
]

designer = [
    "illustrator",
    "oceanographer",
    "medical illustrator",
    "web designer",
    "graphics",
    "designer",
    "editor",
    "film/video"
    "film",
    "animator",
    "artist",
    "theatre director",
    "theatre",
    "art",
    "gallery",
    "architect"
]

food = [
    "restaurant",
    "food",
    "hotel",
    "catering",
    "farm"
]

travel = [
    "tour",
    "tourist",
    "leisure",
    "travel",
    "travel agency",
    "logistics and distribution manager",
    "quality manager",
    "theme park",
    "travel agency"
    
]

manager = [
    "retail manager",
    "customer service manager",
    "information systems manager",
    "product manager",
    "production manager",
    "call centre manager",
    "facilities manager",
    "risk manager",
    "records manager",
    "office manager",
    "location manager",
    "stage manager",
    "conference centre manager",
    "quarry manager",
    "energy manager",
    "television floor manager",
    "forest/woodland manager"
]

real_estate = [
    "housing manager", 
    "estate manager", 
    "purchasing", 
    "warehouse manager", 
    "accommodation manager",
    "public house manager", 
    "heritage manager",
    "logistics",
    "town planner",
    "industrial buyer"
]

entertainment = [
    "actor",
    "radio",
    "social",
    "media"
    
]


def get_customer_job(dataset_row):
    if [keyword for keyword in software if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "software"
        
    elif [keyword for keyword in mathematician if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "mathematician"
    
    elif [keyword for keyword in scientist if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "scientist"
    
    elif [keyword for keyword in teaching if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "teaching"
        
    elif [keyword for keyword in airline if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "airline"
        
    elif [keyword for keyword in medical if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "medicine"
        
    elif [keyword for keyword in marketing if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "marketing"
        
    elif [keyword for keyword in nature if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "nature"
        
    elif [keyword for keyword in trader if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "trader"
        
    elif [keyword for keyword in lawyer if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "lawyer"
        
    elif [keyword for keyword in journalism if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "journalism"
        
    elif [keyword for keyword in finance if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "finance"
        
    elif [keyword for keyword in civil if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "civil"
    
    elif [keyword for keyword in executive if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "executive"
    
    elif [keyword for keyword in media if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "media"
    
    elif [keyword for keyword in sports if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "sports"
        
    elif [keyword for keyword in designer if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "designer"
        
    elif [keyword for keyword in politician if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "politician"
    
    elif [keyword for keyword in food if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "food"
        
    elif [keyword for keyword in other_engineering if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "other_engineering"
    
    elif [keyword for keyword in travel if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "travel"
    
    elif [keyword for keyword in manager if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "manager"
    
    elif [keyword for keyword in real_estate if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "real_estate"
    
    elif [keyword for keyword in entertainment if keyword in dataset_row["customer_job"].lower()]:
        customer_job_capped = "entertainment"
    
    else:
        customer_job_capped = "others"
        
    return customer_job_capped


filtered_dataset["customer_job_capped"] = filtered_dataset.apply(get_customer_job, axis=1)
filtered_dataset.dtypes

event_label                     int64
label_timestamp        datetime64[ns]
event_id                       object
entity_id                      object
customer_job                   object
user_agent                     object
product_category               object
order_price                   float64
merchant                       object
hour_of_day                     int64
minutes_of_day                  int64
time_of_day                    object
browser                        object
device                         object
customer_job_capped            object
dtype: object

In [114]:
filtered_dataset["customer_job_capped"].value_counts()


others               30051
medicine             19327
designer             11882
other_engineering     8830
scientist             7837
teaching              7267
software              4895
civil                 3843
finance               3837
manager               3245
entertainment         1962
executive             1928
real_estate           1663
travel                1503
food                  1407
sports                1309
trader                1256
journalism            1186
media                  920
airline                703
marketing              654
lawyer                 600
nature                 306
mathematician          152
politician             142
Name: customer_job_capped, dtype: int64

In [115]:
len(list(filtered_dataset["customer_job_capped"].unique()))

25

In [116]:
list(filtered_dataset["customer_job_capped"].unique())

['others',
 'other_engineering',
 'medicine',
 'scientist',
 'travel',
 'designer',
 'sports',
 'executive',
 'software',
 'manager',
 'civil',
 'finance',
 'food',
 'teaching',
 'marketing',
 'trader',
 'lawyer',
 'journalism',
 'entertainment',
 'media',
 'real_estate',
 'airline',
 'nature',
 'mathematician',
 'politician']

In [117]:
filtered_dataset.to_csv("./engineered_fraud_data.csv")

SyntaxError: keyword can't be an expression (682853365.py, line 1)